net/graphsync_fetcher.go

```diff
@@ -224,27 +242,57 @@ func (gsf *GraphSyncFetcher) fetchBlocks(ctx context.Context, cids []cid.Cid, ta
 		efsb.Insert("messageReceipts", gsf.ssb.Matcher())
 	}).Node()
 	errChans := make([]<-chan error, 0, len(cids))
-	requestCtx, requestCancel := context.WithTimeout(ctx, requestTimeout)
-	defer requestCancel()
+	requestChans := make([]<-chan graphsync.ResponseProgress, 0, len(cids))
+	cancelFuncs := make([]func(), 0, len(cids))
 	for _, c := range cids {
-		_, errChan := gsf.exchange.Request(requestCtx, targetPeer, cidlink.Link{Cid: c}, selector)
+		requestCtx, requestCancel := context.WithCancel(ctx)
+		defer requestCancel()
+		requestChan, errChan := gsf.exchange.Request(requestCtx, targetPeer, cidlink.Link{Cid: c}, selector)
 		errChans = append(errChans, errChan)
+		requestChans = append(requestChans, requestChan)
+		cancelFuncs = append(cancelFuncs, requestCancel)
 	}
 	// Any of the multiple parallel requests might fail. Wait for all of them to complete, then
 	// return any error (in this case, the last one to be received).
 	var anyError error
-	for _, errChan := range errChans {
-		for err := range errChan {
+	for i, errChan := range errChans {
+		requestChan := requestChans[i]
+		cancelFunc := cancelFuncs[i]
+		err := gsf.consumeResponse(requestChan, errChan, cancelFunc)
+		if err != nil {
 			anyError = err
 		}
 	}
 	return anyError
 }
 
+func (gsf *GraphSyncFetcher) consumeResponse(requestChan <-chan graphsync.ResponseProgress, errChan <-chan error, cancelFunc func()) error {
+	timer := time.NewTimer(gsf.unresponsiveTimeout)
```

Comment:
:disappointed: our clock package doesn't have a timer, just a ticker, which means we can't mock this during testing (there is a [feature request](https://github.com/jonboulle/clockwork/issues/10) open, bit stale). This should be fine, but wanted to flag it as we should try and avoid any direct calls to the time package.

Reply:
got it! So noted... good to know about the no direct use of time thing -- that would make better sense here than the variadic options pattern... but also, a lot of things to change.

Reply:
Blocking: no new dependencies on `time`. Two options for consideration:
- (ab)use a ticker instead that we can fake for testing
- factor out a "progress watchdog" that removes references to time completely while being implementable in terms of a `time.Timer` in production.

Reply:
appreciate this decision and agree. will make changes.